From 5958f4244a92d7179ea8f8776ec1e52fd4d3100e Mon Sep 17 00:00:00 2001
From: Daniel De Graaf <dgdegra@tycho.nsa.gov>
Date: Wed, 28 Feb 2018 17:52:49 -0500
Subject: [PATCH 1/2] Find e820 regions that include the limit

Signed-off-by: Daniel De Graaf <dgdegra@tycho.nsa.gov>
---
 tboot/common/e820.c | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/tboot/common/e820.c b/tboot/common/e820.c
index a42b357..9af4264 100644
--- a/tboot/common/e820.c
+++ b/tboot/common/e820.c
@@ -673,12 +673,17 @@ void get_highest_sized_ram(uint64_t size, uint64_t limit,
             uint64_t base = e820_base_64(entry);
             uint64_t length = e820_length_64(entry);
 
-            /* over 4GB so use the last region that fit */
-            if ( base + length > limit )
+            /* region starts too high */
+            if ( base + size > limit )
                 break;
             if ( size <= length ) {
                 last_fit_base = base;
                 last_fit_size = length;
+                /* region ends too high, crop it */
+                if ( base + length > limit ) {
+                    last_fit_size = limit - base;
+                    break;
+                }
             }
         }
     }
-- 
2.14.3

